[Merged by Bors] - feat(Vandermonde): vandermonde matrices with exceptional rows

apnelson1 opened the pull request:
A Vandermonde matrix is usually taken to have rows of the form `[1, a, a^2, ... a^(n-1)]`, but it is also in some contexts mathematically meaningful to 'projectively' generalize to rows of the form `[b^(n-1),a * b^(n-2), ..., a^(n-1)]`. Up to scalings, this is effectively just allowing new rows parallel to `[0, ..., 0, 1]`. 

We define a new type of matrix `rectVandermonde` allowing for rows of this generalized form with an arbitrary indexing type for the rows, and `projVandermonde`, which is the square matrix version that directly generalizes `vandermonde`. 
We also prove a formula for their determinant, and derive the proof of `det_vandermonde` as a special case.
 
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta commented:
I'll mark this as WIP, presuming that you're going ahead with the changes [discussed on Zulip](https://leanprover.zulipchat.com/#narrow/channel/144837-PR-reviews/topic/.2322377.20Vandermonde.20matrices.20with.20exceptional.20rows)

apnelson1 commented:
> I'll mark this as WIP, presuming that you're going ahead with the changes [discussed on Zulip](https://leanprover.zulipchat.com/#narrow/channel/144837-PR-reviews/topic/.2322377.20Vandermonde.20matrices.20with.20exceptional.20rows)

I've updated the design as per suggestions, and removed the label.

jcommelin commented on Mathlib/LinearAlgebra/Vandermonde.lean:
```diff
@@ -1,45 +1,98 @@
+  simp [projVandermonde, rectVandermonde, vandermonde]
+
+/-- We don't mark this as `@[simp]` because the RHS is not simp-nf,
+and simplifying it RHS gives a bothersome `Nat` subtraction.  -/
```
```suggestion
and simplifying the RHS gives a bothersome `Nat` subtraction.  -/
```

jcommelin reviewed (commented):
Excellent!

bors d+

jcommelin commented:
Thanks :tada:

bors merge
